### ISSUE TITLE
Add public API to inject the local IEclipseContext

### ISSUE THREAD

**Issue by sewe** [eclipse-platform/eclipse.platform#344]
## Use Case

I want to write an Eclipse 4 command handler that wraps/decorates another command handler. To that effect, my handler has `@CanExecute` and `@Execute` methods which invoke the wrapped handler's respective `@CanExecute` and `@Execute` methods using `ContextInjectionFactory.invoke` like so:

```java
@Execute
public void execute(IEclipseContext context) {
    // Decorator functionality goes here

    IEclipseContext localContext = null; // How to obtain this?
    ContextInjectionFactory.invoke(decoratee, Execute.class, context, localContext, null));
}
```

Unfortunately, this looses any "local context" the `ContextInjectionFactory`? invoking the above method might have had.

## Suggested Solution

Allow the local context to be injected as well, e.g., by using a `@Named` annotation:

```java
@Execute
public void execute(IEclipseContext context, @Named("local") IEclipseContext localContext) {
    // Decorator functionality goes here

    ContextInjectionFactory.invoke(decoratee, Execute.class, context, localContext, null));
}
```

## Workarounds

AFAICT, the only way to currently obtain this local context is using an internal API (`org.eclipse.e4.core.internal.contexts.ContextObjectSupplier`):

```java
@Execute
public void execute(IEclipseContext context, ContextObjectSupplier localContextSupplier) {
    // Decorator functionality goes here

    IEclipseContext localContext = localContextSupplier.getContext(); // internal API :-(
    ContextInjectionFactory.invoke(decoratee, Execute.class, context, localContext, null));
}
```

But this is both internal API and also more awkward to use than getting the local context injected directly and not as a supplier.

**Comment by sewe** [comment]
For the record, I initally asked about this functionality [on StackOverflow](https://stackoverflow.com/questions/75335858/how-to-get-the-local-context-injected-in-an-eclipse-4-command-handler), but no-one had any better workaround to offer.

**Comment by laeubi** [comment]
@sewe I think the problem is that there is not *that* local context there could be many of them and even with different content becuase they are local. So usually you need to know what is decorated (how/why?) and then you just inject whatever you want to pass and create a new local context.

In contrast, and if you are in charge of the place where this happens, one can create a child context and use that for injection.

**Comment by sewe** [comment]
@laeubi I realize that the local context may contain whatever makes sense locally and might vary a lot. But that is not really relevant for my use case.

Suppose I want to write a simple no-op wrapper of a command handler that does absolutely nothing (besides adding a few frames on the call stack):

```java
@Execute
public void execute(IEclipseContext context) {
    ContextInjectionFactory.invoke(wrappedHandler, Execute.class, context, null, null)); // Looses information
}
```

Then this is impossible without potentially changing the wrapped handler's behavior; as I have to invoke the `wrappedHandler` with a `null` local context, the "no-op" wrapper _might_ affect the wrapped handler's behavior - in case it depends on the contents of the local context.

So not being able to access the local context in a handler means that I cannot accurately mimic the handler's invocation.

Of course, no-op wrappers are not terribly useful, but decorators are.

Do this explain my use case?

**Comment by laeubi** [comment]
But how do you know the wrapped handler? I don't think that something like this is really possible or supported in a nice way, so if you don't want to alter the behavior this is simply bad as it hides potential other invocations on that handler as well (`@CanExecute` for example).

So probably it would be better to extend the existing handler and overwrite the method you like to "do nothing" with?

**Comment by sewe** [comment]
Of course, I would also need to forward `@CanExecute` invocations to preserve the behavior; those were just omitted from my example snippet (but I would also need to pass a local context to `ContextInjectionFactory.invoke` for them...)

And how to know the handlers I want to wrap? I can get them from the `EModelService` via `MHandlerContainer` and `MHandler`. This is actually the beauty of the Eclipse 4 application model: You can do something like decorating all handlers with, e.g., logging.
